nano/lib/errors.hpp

```diff
@@ -172,3 +186,297 @@ REGISTER_ERROR_CODES (nano, error_common);
+	}
+
+	error (std::error_code code_a)
+	{
+		code = code_a;
+	}
+
+	error (std::string message_a)
+	{
+		code = nano::error_common::generic;
+		message = std::move (message_a);
+	}
+
+	error (std::exception const & exception_a)
+	{
+		code = nano::error_common::exception;
+		message = exception_a.what ();
+	}
+
+	error & operator= (nano::error const & err_a)
+	{
+		code = err_a.code;
+		message = err_a.message;
+		return *this;
+	}
+
+	error & operator= (nano::error && err_a)
+	{
+		code = err_a.code;
+		message = std::move (err_a.message);
+		return *this;
+	}
+
+	/** Assign error code */
+	error & operator= (const std::error_code code_a)
+	{
+		code = code_a;
+		message.clear ();
+		return *this;
+	}
+
+	/** Assign boost error code (as converted to std::error_code) */
+	error & operator= (const boost::system::error_code & code_a)
+	{
+		code = nano::error_conversion::convert (code_a);
+		message.clear ();
+		return *this;
+	}
+
+	/** Assign boost error code (as converted to std::error_code) */
+	error & operator= (const boost::system::errc::errc_t & code_a)
+	{
+		code = nano::error_conversion::convert (boost::system::errc::make_error_code (code_a));
+		message.clear ();
+		return *this;
+	}
+
+	/** Set the error to nano::error_common::generic and the error message to \p message_a */
+	error & operator= (const std::string message_a)
+	{
+		code = nano::error_common::generic;
+		message = std::move (message_a);
+		return *this;
+	}
+
+	/** Set the error to nano::error_common::generic. */
+	error & operator= (bool is_error)
+	{
+		if (is_error)
+		{
+			code = nano::error_common::generic;
+		}
+		message.clear ();
+		return *this;
+	}
+
+	/** Sets the error to nano::error_common::exception and adopts the exception error message. */
+	error & operator= (std::exception const & exception_a)
+	{
+		code = nano::error_common::exception;
+		message = exception_a.what ();
+		return *this;
+	}
+
+	/** Return true if this#error_code equals the parameter */
+	bool operator== (const std::error_code code_a)
+	{
+		return code == code_a;
+	}
+
+	/** Return true if this#error_code equals the parameter */
+	bool operator== (const boost::system::error_code code_a)
+	{
+		return code.value () == code_a.value ();
+	}
+
+	/** Call the function iff the current error is zero */
+	error & then (std::function<nano::error &()> next)
+	{
+		return code ? *this : next ();
+	}
+
+	/** If the current error is one of the listed codes, reset the error code */
+	template <typename... ErrorCode>
+	error & accept (ErrorCode... err)
+	{
+		// Convert variadic arguments to std::error_code
+		auto codes = { std::error_code (err)... };
+		if (std::any_of (codes.begin (), codes.end (), [this, &codes](auto & code_a) { return code == code_a; }))
+		{
+			code.clear ();
+		}
+
+		return *this;
+	}
+
+	/** Implicit error_code conversion */
+	operator std::error_code () const
+	{
+		return code;
+	}
+
+	/** Implicit bool conversion; true if there's an error */
+	operator bool () const
```

Comment:
I couldn't comment on the other response, so have put it here:
> wezrule - Implicit conversions are the root of all evil :)
> cryptocode - Yeah, nano::error to error_code seemed like a good idea though, it's hard to imagine where error-type -> error-type conversion can cause issues, unless I'm missing something
> At least I wanna keep implicit bool conversion for ergonomics just like error_code does - what do you think?

The Implicit bool conversions are more worrisome as it suffers the famous [safe bool problem](https://en.cppreference.com/w/cpp/language/implicit_conversion#The_safe_bool_problem), where it allows further implicit conversion to int, so also allows this:
`error e; `
`int not_a_bool = e;`

You can make it `explicit` to solve this problem though. The standard is moving further away from implicit conversions (enum class is an example) and as "Effective C++ Item 5:  Be wary of user-defined conversion functions" states - 

> Granting compilers license to perform implicit type conversions usually leads to more harm than good, so don't provide conversion functions unless you're sure you want them

Reply:
I agree in general. I made the exception (!) to nano::error  because we use the `if (did_it_work ())` idiom everywhere. But we can always discuss changing that.